kep/61-real-load-aware-scheduling/61-real-load-aware-scheduling.md

```diff
@@ -0,0 +1,608 @@
+the following extended algorithm is proposed that is called when scheduling replicas of a service, replica-set, replication controller, etc.
+
+
+**Algorithm**
+
+
+
+1. For each node, calculate S = (1000*PTS + PTS*BFBP)
+2. Scale each S to be between 0-100
+3. Return the scores
+
+**Algorithm Analysis**
+
+1. The algorithm favors nodes with high PTS scores with 1000*PTS part. The weight 1000 is chosen to give preference to PTS  and to avoid the congestion of scores upon scaling down, my magnifying the scores 
+2. PTS*BFBP penalizes hot nodes with low BFBP scores.
+3. The final scaled-down value deals with congestion, hot node penalty, and favoring of high PTS score nodes
+
+Example: 
+
+
+| Node | PTS | BFBP | Combined Score |
+|------|-----|------|----------------|
+|  N1  | 100 |  35  |       96       |
+|  N2  |  67 |  75  |       67       |
+|  N3  | 100 |  80  |       100      |
+|  N4  |  82 |  90  |       83       |
+|  N5  |  0  |  90  |        0       |
+
+
+
+### Safe Balancing Plugin
+
+
+#### Score plugin
+
+This plugin would extend the Score extension point. K8s scheduler framework calls the Score function for each node separately when scheduling a pod.
+
+
+**Idea**
+
+
+Balancing load based on average would be risky sometimes, as it does not consider the bursty variations. A safe balancing plugin balances not only the average load but also the risk caused by load variations. Suppose we take the mean (M) and standard deviation (V) of all nodes’ utilization into a mu-sigma plot below. In that case, the safe balancing plugin will make placements such that all nodes’ utilization are aligned on the diagonal line, which is V + M = c. Here, c is a constant indicating the overall cluster utilization average plus the standard deviation.
+
+<img src="images/safe-sched-graph.png" alt="safe-sched-graph" width="700" height="400"/>
+
+
+Following is the algorithm:
+
+
+**Algorithm**
+
+
+1. Get the requested resource for the pod to be scheduled as, <img src="https://render.githubusercontent.com/render/math?math=r">.
+2. Get the sliding window average (<img src="https://render.githubusercontent.com/render/math?math=M">) and standard deviation (<img src="https://render.githubusercontent.com/render/math?math=V">) of resource utilization fraction (range from 0 to 1) for all types of resources (CPU, Memory, GPU, etc.) of the current node to be scored. 
+3. Calculate the score of the current node for each type of resource: <img src="https://render.githubusercontent.com/render/math?math=S_i = M %2B r %2B V">
+4. Get a score for each type of resource and bound it to [0,1]: <img src="https://render.githubusercontent.com/render/math?math=S_i = \min(S_i, 1.0)">
+5. Calculate the node priority score per resource as: <img src="https://render.githubusercontent.com/render/math?math=U_i = (1 - S_i) \times MaxPriority ">
+6. Get the final node score as: <img src="https://render.githubusercontent.com/render/math?math=U =\min(U_i)">
```

Comment:
Two questions:

1. In the initial design, we only care about CPU, so we only calculate the score for CPU, or we still calculate the others?
1. Once we consider multiple metrics, can we say we support [DRF](https://amplab.cs.berkeley.edu/publication/dominant-resource-fairness-fair-allocation-of-multiple-resources-types/)?

Reply:
Bump.

Reply:
@wangchen615 ^

Reply:
@Huang-Wei , I think it is an orthogonal question. DRF can be supported in both cases where the real load on CPU/memory is considered or not. New DRF plugins need to be implemented. They can either take the available resources for new requests (in terms of allocatable) or the actual available resources (in terms of actual usage). Right now, we do not have a concrete plan to implement any DRF plugins yet. However, if there are some interests in DRF plugins in the community, we are happy to help. For the DRF plugin, it would be easy to start with the allocatable and then see whether we can adapt to the real load version when the load watcher is ready.